### ISSUE TITLE
[RFC]: Redesigning Persistent Batch in vLLM

### ISSUE THREAD

**Issue by WoosukKwon** [vllm-project/vllm#23446]
### Motivation.

To reduce the overhead of input data construction, vLLM v1 introduced the **persistent batch** technique, inspired by [[LMDeploy](https://github.com/InternLM/lmdeploy)](https://github.com/InternLM/lmdeploy).

## How vLLM's Persistent Batch Works

At a high level, the mechanism operates as follows:

1. Each request’s state is stored in an `InputBatch`, which consists of CPU tensors shaped either as `[num_reqs]` (e.g., sampling temperature) or `[num_reqs, L]` (e.g., block tables).
2. When `N` requests are scheduled, we reorder `InputBatch` so that the first `N` entries correspond to the scheduled requests.
3. Finally, these `N` entries are then copied from CPU to GPU (using regular `cudaMemcpyAsync`).

The key insight is that the set of scheduled requests typically changes only slightly between steps.
Thus, the `InputBatch` from the previous step can be reused with only incremental updates (e.g., appending `new_block_ids` to the block table and reordering a few rows when some requests finish).

## Limitations of Current Implementation

While effective in many cases, vLLM's persistent batch has several drawbacks:

1. **Efficiency depends on minimal changes between steps**—a condition that doesn’t always hold:
   * (while it's not the default case in vLLM) A scheduler may alternate between prefill and decode batches.
   * With pipeline parallelism, each (micro)batch can involve a completely different set of requests, forcing frequent `InputBatch` flushes every step.

2. **Duplicate state storage**: A secondary structure (`CachedRequestState`) is required to hold states for requests that _are unfinished but not scheduled in the current step_.
   * This introduces inefficiency since two copies of the same states must be updated each step.
   * It also complicates the code (two sources of truth).

3. **Complex request reordering**: Some attention implementations require grouping prefill and decode requests separately. To meet this constraint, we need to re-order all the persistent batch, which is complex and error-prone.

### Proposed Change.

To address these issues, I propose a new design based on separating **persistent request states** from **step-level input batches**.

<img width="1040" height="478" alt="Image" src="https://github.com/user-attachments/assets/8082325d-f7b9-48bb-917d-f6494e265ee9" />

### 1. Persistent Request States

We introduce a `RequestStates` class, which holds all persistent request metadata. Most states are stored as (UVA-mapped) **CPU** tensors, with one entry (or row) per request.

```python
class RequestStates:
    def __init__(self, ...):
        self.req_id_to_idx: dict[str, int] = {}
        self.req_idx_to_id: dict[int, str] = {}
        max_cached_reqs = 2 * max_num_reqs
        self.temperature = torch.zeros(max_cached_reqs, device="cpu", pin_memory=True)
        ...
```

Key differences from the old approach:

* **No reordering**: Each request is assigned a fixed index when first scheduled. This index does not change until the request completes.
* **Single source of truth**: `RequestStates` replaces `CachedRequestState`, eliminating duplicated state storage.

---

### 2. Step-Level Input Batches

On each step, a new `InputBatch` is constructed from `RequestStates`.

```python
@dataclass
class InputBatch:
    idx_mapping: torch.Tensor   # batch_idx -> request_idx
    block_table: torch.Tensor
    slot_mapping: torch.Tensor
    sampling_metadata: SamplingMetadata
    ...
```

`InputBatch` contains the **dense GPU tensors** actually fed into the model and sampler.
**Importantly, the construction of `InputBatch` is fast because it happens on GPU**. Given `index_mapping` (determined by the attention backend), we write a custom Triton kernel that gathers entries from `RequestStates` into dense GPU tensors. The kernel can access the CPU tensors in `RequestStates` because they are using UVA.

---

### 3. Special Handling for Block Tables

<img width="1803" height="456" alt="Image" src="https://github.com/user-attachments/assets/f43cb11e-2397-4f28-9ca5-65a02a79ddd9" />

Block tables are treated differently to minimize CPU–GPU transfers:

* Their persistent state is kept on GPU (unlike other states).
* Each step, `new_block_ids` from the scheduler are incrementally applied directly on GPU.
* A custom Triton kernel gathers the final block table for input.
* `slot_mapping` is also built on GPU using the updated block table.

---

## Benefits of the Redesign

* **Simplicity**: Eliminates reordering logic, reducing complexity and sources of bugs.
* **Performance**:
  * Fewer CPU updates to persistent states (no reordering).
  * Reduced CPU–GPU communication by keeping block tables on GPU.
  * Faster input construction via GPU kernels (e.g., slot mapping).

---

Prototype(WIP): https://github.com/vllm-project/vllm/tree/woosuk/input-prep

### CC List.

cc @njhill @LucasWilkinson @zhuohan123 

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

**Comment by simon-mo** [comment]
There are two microbenchmark that can help with this design:
1. How fast can CPU touch the UVA region after GPU read it? If the UVA region is managed well, we don't want a case where "after GPU read, CUDA needs unlock/moves the bytes back to CPU host RAM for edits". Creating false contention. 
2. How fast is the custom triton kernel when you max out the batch size? In particular, how does it scale, what's a good range of <<<num_blocks, num_grids>>>, what's the kernel launch overhead.

**Comment by WoosukKwon** [comment]
@simon-mo Good point. Micro-benchmarks would be helpful although I believe the overheads will be minimal.

> what's the kernel launch overhead.

I'm pretty sure the overhead will be small because we only launch 3 triton kernels (one for sampling metadata, another for block tables, and the last one for slot mapping).

**Comment by LucasWilkinson** [comment]
This makes sense to me, nice proposal! I think

> On each step, a new InputBatch is constructed from RequestStates.

would make async input preparation much easier; which would be super nice!

Noob question (Im only learning about UVA now due to this RFC haha); but with UVA when you hit host memory from the GPU (Triton kernel) the page gets swapped into GMEM right? so the triton kernels would look something like:

Input Prep Kernel: `RAM -> GMEM -gather-> SM (reg/shared) -> GMEM`
BlkTable Prep Kernel: `GMEM -gather-> SM (reg/shared) -> GMEM`

For the "Input Prep Kernel" you have the slow `RAM -> GMEM` access; would this mean it would be hard to saturate `GMEM <> SM` bandwidth? if so would it makes sense to try to overlap this with the "BlkTable Prep Kernel" kernel? by doing either:

1) Launch on different streams with SM controls
2) Do "Input Prep Kernel" on the CPU using custom C++ OP and writing to pinned memory, saving `GMEM <> SM` transfers; i.e. we would do `RAM -gather-> RAM (pinned) -> GMEM`, and we can run "BlkTable Prep Kernel" in parallel? I think since we'd be dealing with static pre-allocated buffers here we could potentially even spawn a new C++ CPU thread and avoid locking the GIL.

Please correct my understanding if its wrong. I think there might be a chance these kernels would be so fast compared to other CPU side prep code that overlap might not even matter.

**Comment by noooop** [comment]
In #23162, I also found that _update_states and _prepare_inputs consume too much CPU time

<img width="1378" height="821" alt="Image" src="https://github.com/user-attachments/assets/6f5f2e20-d9ef-4431-bd09-9be2291ee492" />

I'm glad to hear that these overheads will be reduced.
